File: js/test/hooks.integration.test.mjs
// Guard behavior of the JS SDK against a real local HTTP server.
//
// The other hook tests assert on fields of the captured body they already expect,
// so an encoding the server cannot read looks identical to a correct one. These
// cases run the full transport and compare each captured body with
// `conformance/hooks/request-preflight.json`.

import assert from 'node:assert/strict';
import { createServer } from 'node:http';
import test from 'node:test';
import { Agento11yClient, defaultConfig } from '../.test-dist/index.js';
import { diffJson, loadPreflightRequest, loadResponses, preflightRequest } from './hooksFixtures.mjs';

const hooksEvaluatePath = '/api/v1/hooks:evaluate';
const responses = loadResponses();

/**
 * Starts a local server whose status, body, and delay come from a per-request
 * responder. Same programmable model as `plugins/pi/src/testHttp.ts`, kept local
 * so the SDK suite does not depend on the pi plugin.
 */
async function startHookServer(respond) {
  const state = { requests: [], inFlight: 0, maxInFlight: 0, errors: [] };
  const server = createServer(async (request, response) => {
    const chunks = [];
    for await (const chunk of request) {
      chunks.push(chunk);
    }
    const raw = Buffer.concat(chunks).toString('utf8');
    const call = { path: request.url ?? '', headers: { ...request.headers }, raw, payload: undefined };
    try {
      call.payload = JSON.parse(raw);
    } catch {
      call.payload = undefined;
    }
    state.requests.push(call);
    state.inFlight += 1;
    state.maxInFlight = Math.max(state.maxInFlight, state.inFlight);

    const out = respond(call) ?? {};
    const send = () => {
      state.inFlight -= 1;
      try {
        response.writeHead(out.status ?? 200, { 'content-type': 'application/json' });
        response.end(out.body ?? JSON.stringify({ action: 'allow', evaluations: [] }));
      } catch (error) {
        // A client that aborted on timeout leaves a torn-down response here.
        // Record it: the timeout cases expect it, and every other case asserts
        // this list is empty.
        state.errors.push(`could not write the response: ${String(error)}`);
      }
    };
    response.on('error', (error) => {
      state.errors.push(`response stream error: ${String(error)}`);
    });
    if (out.delayMs !== undefined && out.delayMs > 0) {
      setTimeout(send, out.delayMs).unref();
      return;
    }
    send();
  });

  await new Promise((resolve) => server.listen(0, '127.0.0.1', resolve));
  const { port } = server.address();
  return {
    ...state,
    requests: state.requests,
    errors: state.errors,
    baseUrl: `http://127.0.0.1:${port}`,
    get maxInFlight() {
      return state.maxInFlight;
    },
    async close() {
      server.closeAllConnections?.();
      await new Promise((resolve, reject) => server.close((err) => (err ? reject(err) : resolve())));
    },
  };
}

function respondWith(status, body) {
  return () => ({ status, body });
}

function newClient(options) {
  const warnings = [];
  const client = new Agento11yClient({
    ...defaultConfig(),
    api: { endpoint: options.baseUrl },
    generationExport: {
      ...defaultConfig().generationExport,
      protocol: 'http',
      endpoint: `${options.baseUrl}/api/v1/generations:export`,
      insecure: true,
      auth: options.auth ?? { mode: 'none' },
    },
    hooks: {
      enabled: options.enabled ?? true,
      phases: options.phases ?? ['preflight'],
      timeoutMs: options.timeoutMs ?? 5_000,
      failOpen: options.failOpen ?? true,
    },
    logger: {
      warn: (message) => warnings.push(String(message)),
      error: () => {},
      debug: () => {},
    },
    generationExporter: { exportGenerations: async () => ({ results: [] }), shutdown: async () => {} },
  });
  return { client, warnings };
}

function assertPreflightRequest(server, { expectResponseWritten = true } = {}) {
  assert.equal(server.requests.length, 1, `expected exactly one hook request, got ${server.requests.length}`);
  if (expectResponseWritten) {
    // A server that failed to answer makes the client fail open, and a fail-open
    // assertion passes whether or not the response was ever sent.
    assert.deepEqual(server.errors, [], `the test server could not answer: ${server.errors.join('; ')}`);
  }
  const [call] = server.requests;
  assert.equal(call.path, hooksEvaluatePath);
  assert.equal(call.headers['content-type'], 'application/json');
  const diffs = diffJson(call.payload, loadPreflightRequest());
  assert.deepEqual(diffs, [], `captured request body does not match the shared fixture:\n${diffs.join('\n')}`);
}

function assertFailOpenWarning(warnings, wantSubstring) {
  const matched = warnings.filter((line) => line.includes('allowing request (failOpen)'));
  assert.ok(matched.length > 0, `a swallowed hook failure must be logged, got ${JSON.stringify(warnings)}`);
  if (wantSubstring !== undefined) {
    assert.ok(
      matched.some((line) => line.includes(wantSubstring)),
      `warning does not mention ${wantSubstring}: ${JSON.stringify(matched)}`,
    );
  }
}

for (const failOpen of [true, false]) {
  const policy = failOpen ? 'fail-open' : 'fail-closed';

  test(`allow proceeds under ${policy}`, async () => {
    const server = await startHookServer(respondWith(200, JSON.stringify(responses.allow)));
    const { client } = newClient({ baseUrl: server.baseUrl, failOpen });
    try {
      const response = await client.evaluateHook(preflightRequest());
      assert.equal(response.action, 'allow');
      assert.equal(response.evaluations[0].ruleId, 'pii-detect');
      assertPreflightRequest(server);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });

  test(`deny is enforced under ${policy}`, async () => {
    const server = await startHookServer(respondWith(200, JSON.stringify(responses.deny)));
    const { client } = newClient({ baseUrl: server.baseUrl, failOpen });
    try {
      const response = await client.evaluateHook(preflightRequest());
      assert.equal(response.action, 'deny');
      assert.equal(response.ruleId, 'block-destructive-bash');
      assert.equal(response.reason, 'Bash(*rm*) is not allowed in this environment');
      assertPreflightRequest(server);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });

  test(`an unconfigured phase sends no request under ${policy}`, async () => {
    const server = await startHookServer(respondWith(500, 'should not be called'));
    const { client } = newClient({ baseUrl: server.baseUrl, failOpen, phases: ['postflight'] });
    try {
      const response = await client.evaluateHook(preflightRequest());
      assert.equal(response.action, 'allow');
      assert.equal(server.requests.length, 0);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });

  test(`disabled hooks send no request under ${policy}`, async () => {
    const server = await startHookServer(respondWith(500, 'should not be called'));
    const { client } = newClient({ baseUrl: server.baseUrl, failOpen, enabled: false });
    try {
      const response = await client.evaluateHook(preflightRequest());
      assert.equal(response.action, 'allow');
      assert.equal(server.requests.length, 0);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });
}

for (const status of [429, 503]) {
  test(`status ${status} fails open with a warning`, async () => {
    const server = await startHookServer(respondWith(status, 'upstream unavailable'));
    const { client, warnings } = newClient({ baseUrl: server.baseUrl, failOpen: true });
    try {
      const response = await client.evaluateHook(preflightRequest());
      assert.equal(response.action, 'allow');
      assertFailOpenWarning(warnings, `status ${status}`);
      assertPreflightRequest(server);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });

  test(`status ${status} fails closed`, async () => {
    const server = await startHookServer(respondWith(status, 'upstream unavailable'));
    const { client } = newClient({ baseUrl: server.baseUrl, failOpen: false });
    try {
      await assert.rejects(() => client.evaluateHook(preflightRequest()), /hook evaluation failed: status/);
      assertPreflightRequest(server);
    } finally {
      await client.shutdown();
      await server.close();
    }
  });
}

test('a malformed response body fails open with a warning', async () => {
  const server = await startHookServer(respondWith(200, '{"action": "allow"'));
  const { client, warnings } = newClient({ baseUrl: server.baseUrl, failOpen: true });
  try {
    const response = await client.evaluateHook(preflightRequest());
    assert.equal(response.action, 'allow');
    assertFailOpenWarning(warnings, 'invalid JSON response');
    assertPreflightRequest(server);
  } finally {
    await client.shutdown();
    await server.close();
  }
});

test('a malformed response body fails closed', async () => {
  const server = await startHookServer(respondWith(200, '{"action": "allow"'));
  const { client } = newClient({ baseUrl: server.baseUrl, failOpen: false });
  try {
    await assert.rejects(() => client.evaluateHook(preflightRequest()), /invalid JSON response/);
    assertPreflightRequest(server);
  } finally {
    await client.shutdown();
    await server.close();
  }
});

const slowResponseDelayMs = 2_000;
const slowResponse = () => ({
  status: 200,
  body: JSON.stringify({ action: 'deny', rule_id: 'late' }),
  delayMs: slowResponseDelayMs,
});

test('a response slower than the client timeout fails open with a warning', async () => {
  const server = await startHookServer(slowResponse);
  const { client, warnings } = newClient({ baseUrl: server.baseUrl, failOpen: true, timeoutMs: 250 });
  try {
    const started = Date.now();
    const response = await client.evaluateHook(preflightRequest());
    const elapsed = Date.now() - started;
    assert.equal(response.action, 'allow');
    assert.ok(elapsed < slowResponseDelayMs, `client waited ${elapsed}ms, so it did not enforce its own timeout`);
    assertFailOpenWarning(warnings);
    // The client is gone by the time the delayed write runs, so a write error is
    // the case under test rather than a broken server.
    assertPreflightRequest(server, { expectResponseWritten: false });
  } finally {
    await client.shutdown();
    await server.close();
  }
});

test('a response slower than the client timeout fails closed', async () => {
  const server = await startHookServer(slowResponse);
  const { client } = newClient({ baseUrl: server.baseUrl, failOpen: false, timeoutMs: 250 });
  try {
    const started = Date.now();
    await assert.rejects(
      () => client.evaluateHook(preflightRequest()),
      /hook evaluation failed/,
      'a timeout must surface as a hook transport failure, not as any thrown error',
    );
    const elapsed = Date.now() - started;
    assert.ok(elapsed < slowResponseDelayMs, `client waited ${elapsed}ms, so it did not enforce its own timeout`);
    assertPreflightRequest(server, { expectResponseWritten: false });
  } finally {
    await client.shutdown();
    await server.close();
  }
});

test('configured auth reaches the server', async () => {
  const server = await startHookServer(respondWith(200, JSON.stringify(responses.allow)));
  const { client } = newClient({
    baseUrl: server.baseUrl,
    auth: { mode: 'basic', basicUser: '12345', basicPassword: 'glc-token', tenantId: '12345' },
  });
  try {
    await client.evaluateHook(preflightRequest());
    const [call] = server.requests;
    assert.equal(call.headers.authorization, `Basic ${Buffer.from('12345:glc-token', 'utf8').toString('base64')}`);
    assert.equal(call.headers['x-scope-orgid'], '12345');
    assert.equal(call.headers['x-agento11y-hook-timeout-ms'], '5000');
    assertPreflightRequest(server);
  } finally {
    await client.shutdown();
    await server.close();
  }
});

test('concurrent evaluations are not serialized', async () => {
  // A guard on the request path must not funnel every caller through one
  // connection.
  const server = await startHookServer(() => ({
    status: 200,
    body: JSON.stringify(responses.allow),
    delayMs: 200,
  }));
  const { client } = newClient({ baseUrl: server.baseUrl });
  try {
    const results = await Promise.all([0, 1, 2, 3].map(() => client.evaluateHook(preflightRequest())));
    assert.deepEqual(
      results.map((r) => r.action),
      ['allow', 'allow', 'allow', 'allow'],
    );
    assert.ok(server.maxInFlight > 1, `evaluations were serialized: peak in-flight ${server.maxInFlight}`);
    assert.deepEqual(server.errors, []);
    const fixture = loadPreflightRequest();
    for (const call of server.requests) {
      assert.deepEqual(diffJson(call.payload, fixture), []);
    }
  } finally {
    await client.shutdown();
    await server.close();
  }
});

File: go/agento11y/hooks_integration_test.go
package agento11y

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"testing"
	"time"

	"go.opentelemetry.io/otel/trace/noop"
)

// Guard behavior of the Go SDK against a real local HTTP server. The other hook
// tests decode the captured body back into the SDK's own request type, so an
// encoding the server cannot read round-trips cleanly and looks identical to a
// correct one. These cases run the full transport and compare each captured body
// with conformance/hooks/request-preflight.json instead.

// hookServerResponse is what the responder returns for one hooks:evaluate
// request: the per-request responder model of the JS plugins/pi/src/testHttp.ts
// helper, whose status, body, and delay cover allow, deny, transport failure,
// timeout, and malformed-response cases from one server.
type hookServerResponse struct {
	status int
	body   string
	delay  time.Duration
}

type hookServerRequest struct {
	path    string
	headers http.Header
	raw     string
	payload any
}

type hookTestServer struct {
	*httptest.Server
	mu          sync.Mutex
	requests    []hookServerRequest
	inFlight    int
	maxInFlight int
}

func startHookTestServer(t *testing.T, respond func(hookServerRequest) hookServerResponse) *hookTestServer {
	t.Helper()
	server := &hookTestServer{}
	server.Server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		raw, _ := io.ReadAll(req.Body)
		call := hookServerRequest{
			path:    req.URL.Path,
			headers: req.Header.Clone(),
			raw:     string(raw),
		}
		_ = json.Unmarshal(raw, &call.payload)

		server.mu.Lock()
		server.requests = append(server.requests, call)
		server.inFlight++
		if server.inFlight > server.maxInFlight {
			server.maxInFlight = server.inFlight
		}
		server.mu.Unlock()
		defer func() {
			server.mu.Lock()
			server.inFlight--
			server.mu.Unlock()
		}()

		out := respond(call)
		if out.delay > 0 {
			time.Sleep(out.delay)
		}
		status := out.status
		if status == 0 {
			status = http.StatusOK
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		// The client aborts on timeout, so a delayed write can fail. That is
		// the case under test.
		_, _ = w.Write([]byte(out.body))
	}))
	t.Cleanup(server.Close)
	return server
}

func (s *hookTestServer) calls() []hookServerRequest {
	s.mu.Lock()
	defer s.mu.Unlock()
	return append([]hookServerRequest(nil), s.requests...)
}

func (s *hookTestServer) concurrentPeak() int {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.maxInFlight
}

func respondWith(status int, body string) func(hookServerRequest) hookServerResponse {
	return func(hookServerRequest) hookServerResponse {
		return hookServerResponse{status: status, body: body}
	}
}

type hookIntegrationOptions struct {
	enabled  bool
	failOpen bool
	phases   []HookPhase
	timeout  time.Duration
	auth     AuthConfig
}

func newHookIntegrationClient(t *testing.T, endpoint string, options hookIntegrationOptions) (*Client, *bytes.Buffer) {
	t.Helper()
	logs := &bytes.Buffer{}
	phases := options.phases
	if phases == nil {
		phases = []HookPhase{HookPhasePreflight}
	}
	timeout := options.timeout
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	failOpen := options.failOpen
	client := NewClient(Config{
		Tracer: noop.NewTracerProvider().Tracer("agento11y-go-hooks-integration"),
		Logger: log.New(logs, "", 0),
		GenerationExport: GenerationExportConfig{
			Protocol:        GenerationExportProtocolHTTP,
			Endpoint:        endpoint + "/api/v1/generations:export",
			Insecure:        BoolPtr(true),
			Auth:            options.auth,
			BatchSize:       1,
			FlushInterval:   time.Hour,
			QueueSize:       1,
			MaxRetries:      1,
			InitialBackoff:  time.Millisecond,
			MaxBackoff:      time.Millisecond,
			PayloadMaxBytes: 1 << 20,
		},
		API: APIConfig{Endpoint: endpoint},
		Hooks: HooksConfig{
			Enabled:  options.enabled,
			Phases:   phases,
			Timeout:  timeout,
			FailOpen: &failOpen,
		},
		testGenerationExporter: newNoopGenerationExporter(nil),
		testDisableWorker:      true,
	})
	t.Cleanup(func() { _ = client.Shutdown(context.Background()) })
	return client, logs
}

func hookFixtureResponseBody(t *testing.T, name string) string {
	t.Helper()
	return string(loadHookResponseFixture(t, name))
}

// assertPreflightHookRequest checks the body the SDK actually put on the wire.
func assertPreflightHookRequest(t *testing.T, server *hookTestServer) {
	t.Helper()
	calls := server.calls()
	if len(calls) != 1 {
		t.Fatalf("expected exactly one hook request, got %d", len(calls))
	}
	call := calls[0]
	if call.path != hooksEvaluatePath {
		t.Errorf("unexpected path %q", call.path)
	}
	if got := call.headers.Get("Content-Type"); got != "application/json" {
		t.Errorf("unexpected content type %q", got)
	}
	for _, diff := range diffJSON("", stripEmptyHookMetadata(call.payload), loadHookFixture(t, "request-preflight.json")) {
		t.Errorf("captured request body does not match the shared fixture: %s", diff)
	}
}

func assertFailOpenWarning(t *testing.T, logs *bytes.Buffer, wantSubstring string) {
	t.Helper()
	text := logs.String()
	if !strings.Contains(text, "allowing request (fail_open)") {
		t.Errorf("a swallowed hook failure must be logged, got %q", text)
	}
	if wantSubstring != "" && !strings.Contains(text, wantSubstring) {
		t.Errorf("warning does not mention %q, got %q", wantSubstring, text)
	}
}

func TestHooksIntegrationAllowUnderBothPolicies(t *testing.T) {
	for _, failOpen := range []bool{true, false} {
		t.Run(hookPolicyName(failOpen), func(t *testing.T) {
			server := startHookTestServer(t, respondWith(http.StatusOK, hookFixtureResponseBody(t, "allow")))
			client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true, failOpen: failOpen})

			resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
			if err != nil {
				t.Fatalf("evaluate hook: %v", err)
			}
			if resp.Action != HookActionAllow {
				t.Fatalf("expected allow, got %q", resp.Action)
			}
			if HookDeniedFromResponse(resp) != nil {
				t.Error("allow response must not produce a denial")
			}
			if len(resp.Evaluations) != 1 || resp.Evaluations[0].RuleID != "pii-detect" {
				t.Errorf("unexpected evaluations: %#v", resp.Evaluations)
			}
			assertPreflightHookRequest(t, server)
		})
	}
}

func TestHooksIntegrationDenyUnderBothPolicies(t *testing.T) {
	for _, failOpen := range []bool{true, false} {
		t.Run(hookPolicyName(failOpen), func(t *testing.T) {
			server := startHookTestServer(t, respondWith(http.StatusOK, hookFixtureResponseBody(t, "deny")))
			client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true, failOpen: failOpen})

			resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
			if err != nil {
				t.Fatalf("evaluate hook: %v", err)
			}
			if resp.Action != HookActionDeny {
				t.Fatalf("expected deny, got %q", resp.Action)
			}
			denied := HookDeniedFromResponse(resp)
			if denied == nil {
				t.Fatal("deny response must produce a denial")
			}
			var typed *HookDeniedError
			if !errors.As(denied, &typed) {
				t.Fatalf("unexpected error type %T", denied)
			}
			if typed.RuleID != "block-destructive-bash" {
				t.Errorf("unexpected rule id %q", typed.RuleID)
			}
			if typed.Reason != "Bash(*rm*) is not allowed in this environment" {
				t.Errorf("unexpected reason %q", typed.Reason)
			}
			assertPreflightHookRequest(t, server)
		})
	}
}

func TestHooksIntegrationHTTPErrorStatuses(t *testing.T) {
	for _, status := range []int{http.StatusTooManyRequests, http.StatusServiceUnavailable} {
		t.Run(http.StatusText(status), func(t *testing.T) {
			t.Run("fail-open", func(t *testing.T) {
				server := startHookTestServer(t, respondWith(status, "upstream unavailable"))
				client, logs := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true, failOpen: true})

				resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
				if err != nil {
					t.Fatalf("fail-open must not surface an error: %v", err)
				}
				if resp.Action != HookActionAllow {
					t.Fatalf("expected allow, got %q", resp.Action)
				}
				assertFailOpenWarning(t, logs, fmt.Sprintf("status %d", status))
				assertPreflightHookRequest(t, server)
			})
			t.Run("fail-closed", func(t *testing.T) {
				server := startHookTestServer(t, respondWith(status, "upstream unavailable"))
				client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true})

				_, err := client.EvaluateHook(context.Background(), preflightHookRequest())
				if !errors.Is(err, ErrHookTransportFailed) {
					t.Fatalf("expected a transport error, got %v", err)
				}
				assertPreflightHookRequest(t, server)
			})
		})
	}
}

func TestHooksIntegrationMalformedResponseBody(t *testing.T) {
	const malformed = `{"action": "allow"`

	t.Run("fail-open", func(t *testing.T) {
		server := startHookTestServer(t, respondWith(http.StatusOK, malformed))
		client, logs := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true, failOpen: true})

		resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
		if err != nil {
			t.Fatalf("fail-open must not surface an error: %v", err)
		}
		if resp.Action != HookActionAllow {
			t.Fatalf("expected allow, got %q", resp.Action)
		}
		assertFailOpenWarning(t, logs, "decode hook response")
		assertPreflightHookRequest(t, server)
	})

	t.Run("fail-closed", func(t *testing.T) {
		server := startHookTestServer(t, respondWith(http.StatusOK, malformed))
		client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true})

		_, err := client.EvaluateHook(context.Background(), preflightHookRequest())
		if !errors.Is(err, ErrHookTransportFailed) {
			t.Fatalf("expected a transport error, got %v", err)
		}
		assertPreflightHookRequest(t, server)
	})
}

func TestHooksIntegrationClientTimeout(t *testing.T) {
	const serverDelay = 2 * time.Second
	slow := func(hookServerRequest) hookServerResponse {
		return hookServerResponse{status: http.StatusOK, body: `{"action":"deny","rule_id":"late"}`, delay: serverDelay}
	}

	t.Run("fail-open", func(t *testing.T) {
		server := startHookTestServer(t, slow)
		client, logs := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{
			enabled:  true,
			failOpen: true,
			timeout:  250 * time.Millisecond,
		})

		started := time.Now()
		resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
		elapsed := time.Since(started)
		if err != nil {
			t.Fatalf("fail-open must not surface an error: %v", err)
		}
		if resp.Action != HookActionAllow {
			t.Fatalf("expected allow, got %q", resp.Action)
		}
		if elapsed >= serverDelay {
			t.Errorf("client waited %s, so it did not enforce its own timeout", elapsed)
		}
		assertFailOpenWarning(t, logs, "")
		assertPreflightHookRequest(t, server)
	})

	t.Run("fail-closed", func(t *testing.T) {
		server := startHookTestServer(t, slow)
		client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{
			enabled: true,
			timeout: 250 * time.Millisecond,
		})

		started := time.Now()
		_, err := client.EvaluateHook(context.Background(), preflightHookRequest())
		elapsed := time.Since(started)
		if !errors.Is(err, ErrHookTransportFailed) {
			t.Fatalf("expected a transport error, got %v", err)
		}
		if elapsed >= serverDelay {
			t.Errorf("client waited %s, so it did not enforce its own timeout", elapsed)
		}
		assertPreflightHookRequest(t, server)
	})
}

func TestHooksIntegrationUnconfiguredPhaseSendsNoRequest(t *testing.T) {
	for _, failOpen := range []bool{true, false} {
		t.Run(hookPolicyName(failOpen), func(t *testing.T) {
			server := startHookTestServer(t, respondWith(http.StatusInternalServerError, "should not be called"))
			client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{
				enabled:  true,
				failOpen: failOpen,
				phases:   []HookPhase{HookPhasePostflight},
			})

			resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
			if err != nil {
				t.Fatalf("evaluate hook: %v", err)
			}
			if resp.Action != HookActionAllow {
				t.Fatalf("expected allow, got %q", resp.Action)
			}
			if calls := server.calls(); len(calls) != 0 {
				t.Errorf("expected no hook request, got %d", len(calls))
			}
		})
	}
}

func TestHooksIntegrationDisabledHooksSendNoRequest(t *testing.T) {
	for _, failOpen := range []bool{true, false} {
		t.Run(hookPolicyName(failOpen), func(t *testing.T) {
			server := startHookTestServer(t, respondWith(http.StatusInternalServerError, "should not be called"))
			client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{failOpen: failOpen})

			resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
			if err != nil {
				t.Fatalf("evaluate hook: %v", err)
			}
			if resp.Action != HookActionAllow {
				t.Fatalf("expected allow, got %q", resp.Action)
			}
			if calls := server.calls(); len(calls) != 0 {
				t.Errorf("expected no hook request, got %d", len(calls))
			}
		})
	}
}

func TestHooksIntegrationConfiguredAuthReachesServer(t *testing.T) {
	server := startHookTestServer(t, respondWith(http.StatusOK, hookFixtureResponseBody(t, "allow")))
	client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{
		enabled: true,
		auth: AuthConfig{
			Mode:          ExportAuthModeBasic,
			BasicUser:     "12345",
			BasicPassword: "glc-token",
			TenantID:      "12345",
		},
	})

	if _, err := client.EvaluateHook(context.Background(), preflightHookRequest()); err != nil {
		t.Fatalf("evaluate hook: %v", err)
	}

	calls := server.calls()
	if len(calls) != 1 {
		t.Fatalf("expected one hook request, got %d", len(calls))
	}
	want := "Basic " + base64.StdEncoding.EncodeToString([]byte("12345:glc-token"))
	if got := calls[0].headers.Get("Authorization"); got != want {
		t.Errorf("unexpected authorization header %q", got)
	}
	if got := calls[0].headers.Get("X-Scope-OrgID"); got != "12345" {
		t.Errorf("unexpected tenant header %q", got)
	}
	if got := calls[0].headers.Get(hookTimeoutHeader); got != "5000" {
		t.Errorf("unexpected hook timeout header %q", got)
	}
	assertPreflightHookRequest(t, server)
}

func TestHooksIntegrationConcurrentEvaluations(t *testing.T) {
	// A guard on the request path must not funnel every caller through one
	// connection.
	server := startHookTestServer(t, func(hookServerRequest) hookServerResponse {
		return hookServerResponse{
			status: http.StatusOK,
			body:   hookFixtureResponseBody(t, "allow"),
			delay:  200 * time.Millisecond,
		}
	})
	client, _ := newHookIntegrationClient(t, server.URL, hookIntegrationOptions{enabled: true})

	var wg sync.WaitGroup
	errs := make([]error, 4)
	actions := make([]HookAction, 4)
	for i := range errs {
		wg.Add(1)
		go func(index int) {
			defer wg.Done()
			resp, err := client.EvaluateHook(context.Background(), preflightHookRequest())
			errs[index] = err
			if resp != nil {
				actions[index] = resp.Action
			}
		}(i)
	}
	wg.Wait()

	for i, err := range errs {
		if err != nil {
			t.Fatalf("evaluation %d failed: %v", i, err)
		}
		if actions[i] != HookActionAllow {
			t.Errorf("evaluation %d: expected allow, got %q", i, actions[i])
		}
	}
	if peak := server.concurrentPeak(); peak < 2 {
		t.Errorf("evaluations were serialized: peak in-flight %d", peak)
	}
	fixture := loadHookFixture(t, "request-preflight.json")
	for i, call := range server.calls() {
		for _, diff := range diffJSON("", stripEmptyHookMetadata(call.payload), fixture) {
			t.Errorf("request %d does not match the shared fixture: %s", i, diff)
		}
	}
}

func hookPolicyName(failOpen bool) string {
	if failOpen {
		return "fail-open"
	}
	return "fail-closed"
}
